### PR TITLE
Dataset upgrade logic simplification

### PR DESCRIPTION
**Pull request by gokulavasan**
Build : https://builds.cask.co/browse/CDAP-DUT5651

### CODE REVIEW

**Comment by ecfm** [comment]
Does `UpgradeValueLoader` just catch exception when the store is not initialized, and retry every minute?

**Comment by gokulavasan** [comment]
@maochf UpgradeValueLoader just records an exception if there is any error in trying to read the flag from the table. It doesn't retry the operation until the load operation is invoked again.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/UpgradeValueLoader.java [hunk 1]
```diff
@@ -80,7 +68,8 @@ public void apply() throws Exception {
             if (!row.isEmpty()) {
               byte[] value = row.get(COLUMN);
               ProjectInfo.Version actual = new ProjectInfo.Version(Bytes.toString(value));
-              if (actual.compareTo(EXPECTED_VERSION) >= 0) {
+              LOG.info("Got this value : {}", actual);
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/UpgradeValueLoader.java]: this log message seems a bit too informal

**Comment by anwar6953** [comment]
Could we merge this change to release/4.1? Or do you plan to cherry-pick a subset of these changes to release/4.1 after merging to develop?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java [hunk 2]
```diff
@@ -448,14 +436,33 @@ public void apply() throws Exception {
...
-  // updated asynchronously.
+  // updated using a cache loader with a fixed timeout.
   public boolean isUpgradeComplete() {
     if (upgradeCacheLoader == null) {
       return false;
     }
     return upgradeCacheLoader.getUnchecked(APP_VERSION_UPGRADE_KEY);
   }
 
+  private synchronized boolean executeUpgradeInTransaction(final Table table, final AtomicInteger maxNumberUpgradeRows)
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java]: Why this method need to be synchronized?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/UpgradeValueLoader.java (hunk 1 above)
Comment (gokulavasan) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/UpgradeValueLoader.java]: Oops, should have removed this debug statement.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java (hunk 2 above)
Comment (gokulavasan) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java]: Is the TransactionExecutorFactory thread safe? The other methods that were accessing it were synchronized.

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java]: The factory should be thread safe.

**Comment by gokulavasan** [comment]
@maochf Please take a look again. I have addressed all the comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/UpgradeHttpHandler.java [hunk 3]
```diff
@@ -33,7 +32,6 @@
 /**
  * {@link HttpHandler} for getting info about upgrade progress, if applicable.
  */
-@Singleton
```
Comment (ecfm) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/UpgradeHttpHandler.java]: Why is this removed?

**Comment by gokulavasan** [comment]
@maochf Addressed comment. Take a look again. Thanks!

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java [hunk 4]
```diff
@@ -117,15 +117,12 @@
...
   private static final Type STRING_MAP_TYPE = new TypeToken<Map<String, String>>() { }.getType();
-  private static final AtomicBoolean appMetaStoreInitialized = new AtomicBoolean(false);
-  private static final AtomicBoolean upgradeComplete = new AtomicBoolean(false);
-  private static final AtomicBoolean cacheLoaderInitialized = new AtomicBoolean(false);
 
   private final CConfiguration configuration;
   private final DatasetFramework dsFramework;
   private final Transactional transactional;
-
-  private static LoadingCache<byte[], Boolean> upgradeCacheLoader;
```
Comment (ecfm) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:128]: what's the motivation for this change?

Comment (gokulavasan) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:128]: It was simpler to just not use a static variable and a separate AtomicBoolean to initialize it.

**Comment by ecfm** [comment]
LGTM
